File: 파이썬/3.스택,큐/9012.py
'''
 * Author : JC
 * https://www.acmicpc.net/problem/9012
'''

t = int(input())

for i in range(t) :
    m = input()
    length = len(m)
    tmp = []
    check = True
    for j in range(length) : 
        if m[j] == '(' :
            tmp.append(1)
        else :
            if len(tmp) == 0 :
                check = False
                break
            else :
                tmp.pop()

    if check == True and len(tmp) == 0: 
        print("YES")
    else :
        print('NO')

File: C++/7. 스택 - 초급/9012.cpp
/*
 * Author : JC
 * https://www.acmicpc.net/problem/9012
*/

#include <iostream>
#include <stack>

using namespace std;

int n;
string str;

int main(){

    cin >> n;
    
    for(int i=0; i<n; i++){
        stack<string> s;
        cin >> str;
        int len = str.length();
        bool check = true;
        for(int j=0; j<len; j++){
            if(str[j] == '('){
                s.push("1");
            }else{
                if(s.empty()){
                    cout << "NO\n";
                    check = false;
                    break;
                }
                s.pop();
            }
        }
        if(check == false){
            continue;
        }
        if(s.empty()){
            cout << "YES\n";
        }else{
            cout << "NO\n";
        }
    }
    
    return 0;
}
